lib/docker.ml

```diff
@@ -1,37 +1,344 @@
+open Lwt.Syntax
+
+type ids = [
+  | `Docker_image of string | `Docker_container of string
+  | `Docker_volume of string | `Obuilder_id of string
+]
+
+let prefix = ref "obuilder"
+let set_prefix prefix' = prefix := prefix'
+
+let image_prefix () = !prefix ^ "-image-"
+let container_prefix () = !prefix ^ "-container-"
+let cache_prefix () = !prefix ^ "-cache-"
+let volume_prefix () = !prefix ^ "-copy-"
+
+let obuilder_volume () = !prefix ^ "-volume"
+let image_name ?(tmp=false) name = image_prefix () ^ (if tmp then "tmp-" else "") ^ name
+let container_name name = container_prefix () ^ name
+let volume_cache_name ?(tmp=false) name = cache_prefix () ^ (if tmp then "tmp-" else "") ^ name
+let volume_copy_name ?(tmp=false) name = volume_prefix () ^ (if tmp then "tmp-" else "") ^ name
+
+let docker_image ?(tmp=false) id = `Docker_image (image_name ~tmp id)
+let docker_container id = `Docker_container (container_name id)
+let docker_volume_cache ?(tmp=false) id = `Docker_volume (volume_cache_name ~tmp id)
+let docker_volume_copy ?(tmp=false) id = `Docker_volume (volume_copy_name ~tmp id)
+
+let ( / ) = Filename.concat
+let mount_point_inside_unix = if Sys.win32 then "/cygdrive/c" else "/var/lib/obuilder"
+let mount_point_inside_native = if Sys.win32 then {|C:/|} else mount_point_inside_unix
+
+let bash_entrypoint obuilder_volume =
+  [if Sys.win32 then mount_point_inside_native / obuilder_volume / "bash.exe" else "bash"; "-c"]
+
+let default_entrypoint =
+  if Sys.win32 then [{|C:\Windows\System32\cmd.exe|}; "/S"; "/C"]
+  else ["/bin/sh"; "-c"]
+
+let rec setup_command ~entp ~cmd =
+  match entp with
+  | hd :: tl -> hd, tl @ cmd
+  | [] -> setup_command ~entp:default_entrypoint ~cmd
+
+let extract_name = function `Docker_image name | `Docker_container name | `Docker_volume name -> name
+
+let pread ?stderr argv =
+  let stderr = Option.value ~default:(`FD_move_safely Os.stderr) stderr in
+  Os.pread ~stderr ("docker" :: argv)
+
+let pread_result ?stdin ?stderr argv =
+  let cmd = "docker" :: argv in
+  let pp f = Os.pp_cmd f ("", cmd) in
+  let stdin = Option.value ~default:`Dev_null stdin in
+  let stderr = Option.value ~default:(`FD_move_safely Os.stderr) stderr in
+  Os.pread_result ~pp ~stdin ~stderr cmd
+
+let exec' ?stdin ?stdout ?stderr ?is_success argv =
+  let stdin = Option.value ~default:`Dev_null stdin in
+  let stdout = Option.value ~default:(`FD_move_safely Os.stdout) stdout in
+  let stderr = Option.value ~default:(`FD_move_safely Os.stderr) stderr in
+  Os.exec ~stdin ~stdout ~stderr ?is_success ("docker" :: argv)
+
+let exec_result' ?stdin ?stdout ?stderr ?is_success argv =
+  let cmd = "docker" :: argv in
+  let pp f = Os.pp_cmd f ("", cmd) in
+  let stdin = Option.value ~default:`Dev_null stdin in
+  let stdout = Option.value ~default:(`FD_move_safely Os.stdout) stdout in
+  let stderr = Option.value ~default:(`FD_move_safely Os.stderr) stderr in
+  Os.exec_result ~stdin ~stdout ~stderr ?is_success ~pp cmd
+
+module Cmd = struct
```

Comment:
Good to have integration tests for these commands. Run against docker linux, macos and windows.